In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from IPython.display import Image

In [2]:
df = pd.read_csv('mkt_data.csv')

In [3]:
df.head()

Unnamed: 0   Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0           0  58138.0        0         0       58       635         88   
1           1  46344.0        1         1       38        11          1   
2           2  71613.0        0         0       26       426         49   
3           3  26646.0        1         0       26        11          4   
4           4  58293.0        1         0       94       173         43   

   MntMeatProducts  MntFishProducts  MntSweetProducts  ...  \
0              546              172                88  ...   
1                6                2                 1  ...   
2              127              111                21  ...   
3               20               10                 3  ...   
4              118               46                27  ...   

   education_Graduation  education_Master  education_PhD  MntTotal  \
0                   3.0               NaN            NaN      1529   
1                   3.0               NaN            NaN        21   
2                   3.0               NaN            NaN       734   
3                   3.0               NaN            NaN        48   
4                   NaN               NaN            5.0       407   

   MntRegularProds  AcceptedCmpOverall  marital_status  education_level  kids  \
0             1441                   0          Single       Graduation     0   
1               15                   0          Single       Graduation     2   
2              692                   0        Together       Graduation     0   
3               43                   0        Together       Graduation     1   
4              392                   0         Married              PhD     1   

   expenses  
0      1529  
1        21  
2       734  
3        48  
4       407  

[5 rows x 44 columns]

- How much data do we have? Rows and columns
- What are the numerical columns?
- Do we have duplicates in our dataset?
- Do we have null values in this dataset? Could they indicate something? What should we do with them?
- What are the mean, median, 25th percentile, 75th percentile, minimum, and maximum of each numerical column?

In [4]:
df.shape

(2205, 44)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2205 entries, 0 to 2204
Data columns (total 44 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            2205 non-null   int64  
 1   Income                2205 non-null   float64
 2   Kidhome               2205 non-null   int64  
 3   Teenhome              2205 non-null   int64  
 4   Recency               2205 non-null   int64  
 5   MntWines              2205 non-null   int64  
 6   MntFruits             2205 non-null   int64  
 7   MntMeatProducts       2205 non-null   int64  
 8   MntFishProducts       2205 non-null   int64  
 9   MntSweetProducts      2205 non-null   int64  
 10  MntGoldProds          2205 non-null   int64  
 11  NumDealsPurchases     2205 non-null   int64  
 12  NumWebPurchases       2205 non-null   int64  
 13  NumCatalogPurchases   2205 non-null   int64  
 14  NumStorePurchases     2205 non-null   int64  
 15  NumWebVisitsMonth    

In [6]:
df = df.drop('Unnamed: 0', axis=1)

In [7]:
df.duplicated().sum()

np.int64(184)

In [8]:
df = df.drop_duplicates()

In [9]:
df.isnull().sum()

Income                     0
Kidhome                    0
Teenhome                   0
Recency                    0
MntWines                   0
MntFruits                  0
MntMeatProducts            0
MntFishProducts            0
MntSweetProducts           0
MntGoldProds               0
NumDealsPurchases          0
NumWebPurchases            0
NumCatalogPurchases        0
NumStorePurchases          0
NumWebVisitsMonth          0
AcceptedCmp3               0
AcceptedCmp4               0
AcceptedCmp5               0
AcceptedCmp1               0
AcceptedCmp2               0
Complain                   0
Z_CostContact              0
Z_Revenue                  0
Response                   0
Age                        0
Customer_Days              0
marital_Divorced        1807
marital_Married         1236
marital_Single          1578
marital_Together        1512
marital_Widow           1951
education_2n Cycle      1838
education_Basic         1972
education_Graduation    1006
education_Mast

The null values mentioned above may indicate that the person does not have that particular characteristic. Note that these columns only have null values or one other specific value. For example:

In [10]:
df['marital_Single'].value_counts()


marital_Single
4.0    443
Name: count, dtype: int64

The column "marital_Single" only contains the number 4.0, which appears 443 times. The rest are null values. This is a strong indication that the null values mean the person does not have that specific characteristic.

Therefore, these columns can be converted into a boolean format — 1 if true and 0 if false (null).

In [11]:
null_columns =["marital_Divorced"
, "marital_Married"
, "marital_Single"
, "marital_Together"
, "marital_Widow"
, "education_2n Cycle"
, "education_Basic"
, "education_Graduation"
, "education_Master"
, "education_PhD"]

for n in null_columns:
    df["booleano "+str(n)] = np.where(df[n].isnull(), 0, 1)

In [12]:
df

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0     58138.0        0         0       58       635         88   
1     46344.0        1         1       38        11          1   
2     71613.0        0         0       26       426         49   
3     26646.0        1         0       26        11          4   
4     58293.0        1         0       94       173         43   
...       ...      ...       ...      ...       ...        ...   
2198  26816.0        0         0       50         5          1   
2200  61223.0        0         1       46       709         43   
2202  56981.0        0         0       91       908         48   
2203  69245.0        0         1        8       428         30   
2204  52869.0        1         1       40        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  ...  \
0                 546              172                88            88  ...   
1                   6                2                 1             6  ...   
2                 127              111                21            42  ...   
3                  20               10                 3             5  ...   
4                 118               46                27            15  ...   
...               ...              ...               ...           ...  ...   
2198                6                3                 4             3  ...   
2200              182               42               118           247  ...   
2202              217               32                12            24  ...   
2203              214               80                30            61  ...   
2204               61                2                 1            21  ...   

      booleano marital_Divorced  booleano marital_Married  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         1   
...                         ...                       ...   
2198                          0                         0   
2200                          0                         1   
2202                          1                         0   
2203                          0                         0   
2204                          0                         1   

      booleano marital_Single  booleano marital_Together  \
0                           1                          0   
1                           1                          0   
2                           0                          1   
3                           0                          1   
4                           0                          0   
...                       ...                        ...   
2198                        1                          0   
2200                        0                          0   
2202                        0                          0   
2203                        0                          1   
2204                        0                          0   

      booleano marital_Widow  booleano education_2n Cycle  \
0                          0                            0   
1                          0                            0   
2                          0                            0   
3                          0                            0   
4                          0                            0   
...                      ...                          ...   
2198                       0                            0   
2200                       0                            0   
2202                       0                            0   
2203                       0                            0   
2204                       0                            0   

      booleano education_Basic  booleano education_Graduation  \
0            

In [13]:
df.describe()

Income      Kidhome     Teenhome      Recency     MntWines  \
count    2021.000000  2021.000000  2021.000000  2021.000000  2021.000000   
mean    51687.258783     0.443345     0.509649    48.880752   306.492331   
std     20713.046401     0.536196     0.546393    28.950917   337.603877   
min      1730.000000     0.000000     0.000000     0.000000     0.000000   
25%     35416.000000     0.000000     0.000000    24.000000    24.000000   
50%     51412.000000     0.000000     0.000000    49.000000   178.000000   
75%     68274.000000     1.000000     1.000000    74.000000   507.000000   
max    113734.000000     2.000000     2.000000    99.000000  1493.000000   

         MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
count  2021.000000      2021.000000      2021.000000       2021.000000   
mean     26.364671       166.059871        37.603662         27.268679   
std      39.776518       219.869126        54.892196         41.575454   
min       0.000000         0.000000         0.000000          0.000000   
25%       2.000000        16.000000         3.000000          1.000000   
50%       8.000000        68.000000        12.000000          8.000000   
75%      33.000000       230.000000        50.000000         34.000000   
max     199.000000      1725.000000       259.000000        262.000000   

       MntGoldProds  ...  booleano marital_Divorced  booleano marital_Married  \
count   2021.000000  ...                2021.000000               2021.000000   
mean      43.921821  ...                   0.105888                  0.388422   
std       51.678211  ...                   0.307771                  0.487512   
min        0.000000  ...                   0.000000                  0.000000   
25%        9.000000  ...                   0.000000                  0.000000   
50%       25.000000  ...                   0.000000                  0.000000   
75%       56.000000  ...                   0.000000                  1.000000   
max      321.000000  ...                   1.000000                  1.000000   

       booleano marital_Single  booleano marital_Together  \
count              2021.000000                2021.000000   
mean                  0.219198                   0.251856   
std                   0.413806                   0.434186   
min                   0.000000                   0.000000   
25%                   0.000000                   0.000000   
50%                   0.000000                   0.000000   
75%                   0.000000                   1.000000   
max                   1.000000                   1.000000   

       booleano marital_Widow  booleano education_2n Cycle  \
count             2021.000000                  2021.000000   
mean                 0.034636                     0.090549   
std                  0.182902                     0.287038   
min                  0.000000                     0.000000   
25%                  0.000000                     0.000000   
50%                  0.000000                     0.000000   
75%                  0.000000                     0.000000   
max                  1.000000                     1.000000   

       booleano education_Basic  booleano education_Graduation  \
count               2021.000000                    2021.000000   
mean                   0.024245                       0.502227   
std                    0.153848                       0.500119   
min                    0.000000                       0.000000   
25%                    0.000000                       0.000000   
50%                    0.000000                       1.000000   
75%                    0.000000                       1.000000   
max                    1.000000                       1.000000   

       booleano education_Master  booleano education_PhD  
count                2021.000000             2021.000000  
mean                    0.165760                0.217219  
std                     0.371957                0.4

Understanding our customer profile better:

- What is the highest salary found in our dataset?
- What is the salary distribution in our dataset?
- Do our customers have higher or lower education levels?
- How many customers do we have in each marital statu

In [14]:
df['Income'].max()

np.float64(113734.0)

In [37]:
fig = px.histogram(data_frame=df, x='Income', title='Income Distribution')
fig.update_layout(bargap=0.1)
fig.update_layout(xaxis_title=None)
fig.write_image('scatter_plot.png')
img = fig.to_image(format='png')
Image(img)
fig


In [36]:
fig = px.box(df, y="Income", title="Income Distribution")
fig.write_image('box_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

In [17]:
# Coefficient of variation (CV)
df['Income'].std()/df['Income'].mean()


np.float64(0.4007379553227728)

In [18]:
df['Income'].skew()

np.float64(0.004357040710783565)

Our dataset shows well-divided salaries. A few people earn little, and a few earn a lot, but there are no outliers.

The dispersion, on the other hand, is large (many data points deviate from the mean).

Based on the skewness and the boxplot, we can say that, although not completely normal, the curve is close to a normal distribution.

In [35]:
fig = px.bar(df, x='education_level', title='Education Level')
fig.update_layout(xaxis_title=None)
fig.update_traces(marker_color='blue')
fig.write_image('bar_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

Most of our clients have completed a bachelor's degree. Few clients have only completed basic education.
In general, we can say that our clients have higher levels of education.

In [34]:
fig = px.bar(df, x='marital_status', title='Marital Status')
fig.update_layout(xaxis_title=None)
fig.write_image('bar_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

How the data correlates:

- What is the relationship between marital status and number of children? Do married people have more children?
- Do people spend more or less on our platform when they have children?
- Do people with higher salaries spend more?


In [33]:
fig = px.box(df, x='marital_status', y='kids')
fig.write_image('box_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

In [22]:
df.groupby('marital_status')['kids'].describe()

count      mean       std  min  25%  50%  75%  max
marital_status                                                    
Divorced        214.0  1.023364  0.753344  0.0  1.0  1.0  1.0  3.0
Married         785.0  0.977070  0.724558  0.0  0.0  1.0  1.0  3.0
Single          443.0  0.860045  0.771850  0.0  0.0  1.0  1.0  3.0
Together        509.0  0.982318  0.748806  0.0  0.0  1.0  1.0  3.0
Widow            70.0  0.842857  0.734960  0.0  0.0  1.0  1.0  2.0

In [23]:
df.groupby('marital_status')['kids'].median()

marital_status
Divorced    1.0
Married     1.0
Single      1.0
Together    1.0
Widow       1.0
Name: kids, dtype: float64

Although the numbers are very close, looking only at the average, we can say that singles have fewer children compared to the other groups.
However, as we can see in the boxplot and the median, the numbers are very close to each other.
We would need a hypothesis test to confirm any differences.

In [32]:
fig = px.box(df, x='kids', y='expenses')
fig.write_image('box_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

In [25]:
df.groupby('kids')['expenses'].describe()

count         mean         std  min    25%     50%      75%     max
kids                                                                      
0      567.0  1056.282187  622.282197  5.0  605.5  1114.0  1531.50  2491.0
1     1027.0   432.732230  450.304666  4.0   50.5   265.0   735.50  2036.0
2      382.0   223.267016  295.568285  8.0   39.0    84.0   295.75  1627.0
3       45.0   240.044444  407.638708  6.0   19.0    62.0   254.00  1601.0

In [26]:
df.groupby('kids')['expenses'].median()

kids
0    1114.0
1     265.0
2      84.0
3      62.0
Name: expenses, dtype: float64


We can see that both the average and the median for 0 and 1 child are higher in this data sample. However, the average for 2 children is lower than for 3 children, while the median for 2 children is higher than for 3 children.
Both the 2-child and 3-child groups have averages much higher than the median, indicating upper outliers (which the graph itself already shows). The 1-child group also has upper outliers, but this difference is less pronounced.
People with no children have a distribution that is somewhat more similar to a normal distribution (no outliers, and the median is slightly centered in the middle of the boxplot).

In [31]:
fig = px.scatter(df, x='expenses', y='Income')
fig.write_image('scatter_plot.png')
img = fig.to_image(format='png')
Image(img)
fig

In [28]:
df["expenses"].corr(df["Income"])

np.float64(0.8192029927546332)

Apparently, there is a directly proportional correlation, where people who earn more also spend more.